File: frontend/types/alert.ts
// Alert domain types
// AUTO-SYNCED with backend via openapi-typescript
// Last synced: 2026-01-08
// Source: components["schemas"]["AlertConfigurationCreate"], AlertRead, etc.

// ============================================
// ENUMS / UNION TYPES
// ============================================

/**
 * Alert severity levels
 * Matches: components["schemas"]["AlertSeverity"]
 */
export type AlertSeverity = 'low' | 'medium' | 'high' | 'critical';

/**
 * Alert lifecycle status
 * Matches: components["schemas"]["AlertStatus"]
 */
export type AlertStatus = 'pending' | 'sent' | 'failed' | 'acknowledged' | 'resolved';

/**
 * Types of alerts the system can generate
 * Matches: components["schemas"]["AlertType"]
 */
export type AlertType =
  | 'sentiment_drop'
  | 'sentiment_spike'
  | 'volume_surge'
  | 'viral_mention'
  | 'competitor_price_change'
  | 'price_recommendation'
  | 'price_applied'
  | 'trend_detected'
  | 'anomaly_detected';

/**
 * Notification delivery channels
 * Matches: components["schemas"]["AlertChannel"]
 */
export type AlertChannel = 'email' | 'slack' | 'webhook' | 'in_app';

// ============================================
// ALERT TYPES
// ============================================

/**
 * Alert response from GET endpoints
 * Matches: components["schemas"]["AlertRead"]
 */
export interface Alert {
  id: string;
  user_id: string;
  configuration_id: string | null;
  alert_type: AlertType;
  severity: AlertSeverity;
  title: string;
  message: string;
  product_id: string | null;
  competitor_id: string | null;
  recommendation_id: string | null;
  data: Record<string, unknown>;
  status: AlertStatus;
  channels_sent: string[];
  channels_failed: string[];
  created_at: string;
  sent_at: string | null;
  acknowledged_at: string | null;
  acknowledged_by: string | null;
  resolved_at: string | null;
}

/**
 * Alert stats for dashboard
 * Matches: components["schemas"]["AlertStats"]
 */
export interface AlertStats {
  total_unread: number;
  by_severity: Record<string, number>;
  by_type: Record<string, number>;
  recent_24h: number;
}

/**
 * Paginated alerts response
 */
export interface PaginatedAlerts {
  items: Alert[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

/**
 * Alert filter params for list endpoint
 */
export interface AlertFilterParams {
  skip?: number;
  limit?: number;
  status?: AlertStatus;
  severity?: AlertSeverity;
  alert_type?: AlertType;
  product_id?: string;
}

// ============================================
// ALERT CONFIGURATION TYPES
// ============================================

/**
 * Alert configuration response
 * Matches: components["schemas"]["AlertConfigurationRead"]
 */
export interface AlertConfiguration {
  id: string;
  user_id: string;
  name: string;
  description: string | null;
  alert_type: AlertType;
  is_active: boolean;
  product_ids: string[] | null;
  conditions: Record<string, unknown>;
  channels: AlertChannel[];
  channel_settings: Record<string, unknown>;
  cooldown_minutes: number;
  max_per_day: number;
  created_at: string;
  updated_at: string;
  last_triggered_at: string | null;
}

/**
 * Create alert configuration request
 * Matches: components["schemas"]["AlertConfigurationCreate"]
 */
export interface AlertConfigurationCreate {
  name: string;
  description?: string | null;
  alert_type: AlertType;
  is_active?: boolean;                    // Default: true
  product_ids?: string[] | null;
  conditions?: Record<string, unknown>;
  channels?: AlertChannel[];              // Default: ["in_app"]
  channel_settings?: Record<string, unknown>;
  cooldown_minutes?: number;              // Default: 60
  max_per_day?: number;                   // Default: 10
}

/**
 * Update alert configuration request
 * Matches: components["schemas"]["AlertConfigurationUpdate"]
 */
export interface AlertConfigurationUpdate {
  name?: string | null;
  description?: string | null;
  is_active?: boolean | null;
  product_ids?: string[] | null;
  conditions?: Record<string, unknown> | null;
  channels?: AlertChannel[] | null;
  channel_settings?: Record<string, unknown> | null;
  cooldown_minutes?: number | null;
  max_per_day?: number | null;
}

// AlertAnalytics is defined in types/analytics.ts and exported from there

File: backend/schemas/alert.py
# backend/schemas/alert.py
"""Alert schemas for API request/response validation."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from models.alert import AlertChannel, AlertSeverity, AlertStatus, AlertType

# ============== AlertConfiguration Schemas ==============


class AlertConfigurationCreate(BaseModel):
    """Create a new alert configuration."""

    name: str = Field(..., max_length=255)
    description: str | None = None
    alert_type: AlertType
    is_active: bool = True
    product_ids: list[UUID] | None = None
    conditions: dict[str, Any] = Field(default_factory=dict)
    channels: list[AlertChannel] = Field(default=[AlertChannel.IN_APP])
    channel_settings: dict[str, Any] = Field(default_factory=dict)
    cooldown_minutes: int = Field(default=60, ge=1)
    max_per_day: int = Field(default=10, ge=1)


class AlertConfigurationUpdate(BaseModel):
    """Update an existing alert configuration."""

    name: str | None = Field(None, max_length=255)
    description: str | None = None
    is_active: bool | None = None
    product_ids: list[UUID] | None = None
    conditions: dict[str, Any] | None = None
    channels: list[AlertChannel] | None = None
    channel_settings: dict[str, Any] | None = None
    cooldown_minutes: int | None = Field(None, ge=1)
    max_per_day: int | None = Field(None, ge=1)


class AlertConfigurationRead(BaseModel):
    """Alert configuration response."""

    model_config = ConfigDict(from_attributes=True)
    id: UUID
    user_id: UUID
    name: str
    description: str | None
    alert_type: AlertType
    is_active: bool
    product_ids: list[UUID] | None
    conditions: dict[str, Any]
    channels: list[AlertChannel]
    channel_settings: dict[str, Any]
    cooldown_minutes: int
    max_per_day: int
    created_at: datetime
    updated_at: datetime
    last_triggered_at: datetime | None


# ============== Alert Schemas ==============


class AlertCreate(BaseModel):
    """Create a new alert (usually system-generated)."""

    alert_type: AlertType
    severity: AlertSeverity = AlertSeverity.MEDIUM
    title: str = Field(..., max_length=255)
    message: str
    product_id: UUID | None = None
    competitor_id: UUID | None = None
    recommendation_id: UUID | None = None
    configuration_id: UUID | None = None
    data: dict[str, Any] = Field(default_factory=dict)


class AlertRead(BaseModel):
    """Alert response."""

    model_config = ConfigDict(from_attributes=True)
    id: UUID
    user_id: UUID
    configuration_id: UUID | None
    alert_type: AlertType
    severity: AlertSeverity
    title: str
    message: str
    product_id: UUID | None
    competitor_id: UUID | None
    recommendation_id: UUID | None
    data: dict[str, Any]
    status: AlertStatus
    channels_sent: list[str]
    channels_failed: list[str]
    created_at: datetime
    sent_at: datetime | None
    acknowledged_at: datetime | None
    acknowledged_by: UUID | None
    resolved_at: datetime | None


class AlertAcknowledge(BaseModel):
    """Acknowledge an alert."""

    pass  # No body needed, user comes from auth


class AlertResolve(BaseModel):
    """Resolve an alert."""

    resolution_note: str | None = None


# ============== Stats/Summary Schemas ==============


class AlertStats(BaseModel):
    """Alert statistics for dashboard."""

    total_unread: int
    by_severity: dict[str, int]  # {"critical": 2, "high": 5, ...}
    by_type: dict[str, int]  # {"sentiment_drop": 3, ...}
    recent_24h: int


class AlertListResponse(BaseModel):
    """Paginated alert list response."""

    alerts: list[AlertRead]
    total: int
    page: int
    per_page: int
    has_more: bool
